sdk/client.py

```diff
@@ -146,21 +41,66 @@ def __init__(
         password: Optional[str] = None,
         token_provider: Optional[TokenProvider] = None,
         endpoints: Optional[Endpoints] = None,
+        auth0: Optional[Auth0Conf] = None,
     ):
-        if not (username and password) and not token_provider:
-            raise Exception(
-                "At least a username/password combination or"
-                "TokenProvider object should be provided."
+        if not username and not token_provider:
+            raise ValueError(
+                "At least a username or TokenProvider object should be provided."
             )
-        self.endpoints = endpoints or Endpoints()
-        self.group_id = group_id
+        if token_provider is not None:
+            self._check_token_provider(token_provider)
 
-        if username and password:
-            token_provider = UsernamePasswordTokenProvider(
-                username, password, f"{self.endpoints.account}/api/v1/auth/login"
-            )
+        self.endpoints = self._make_endpoints(endpoints)
+
+        if username:
+            auth0 = self._make_auth0(auth0)
+            token_provider = self._credential_login(username, password, auth0)
 
         self.authenticator = HTTPBearerAuthenticator(token_provider)
+        self.group_id = group_id
```

Comment:
nit: I feel like this is a bit inconsistent in terms of naming and implementation. 
Some of these are "check", some are "make", there is "credential_login" which builds a TokenProvider and does not login. Could we try to uniformize it a bit and define clearly what is the goal of these functions (user input validation, set defaults, .. ) ?

Reply:
To me, it's crystal clear, but maybe I was too much into the feature myself to realize it's not clear.
The `_make_xxx` is used because we have a parameter, and it's used to assure yourself that the returned value is valid for your implementation.
The `_credential_login` actually makes a login request, and therefore gets a `token` that will be used in the `HTTPBearerAuthenticator`.
The `_check_...` does not return something and only check if the value is correct.

Maybe we want to make a convention in order to uniformize this ?

Reply:
It's fine